### PR TITLE
[mysql] add validate method verify all startup modes whether the parameter is legal

### LINKED ISSUE #1743: FlinkCDC_2.3.0 TableAPI didn't check "scan.startup.mode" when the mode is "timestamp"
**Describe the bug(Please use English)**
使用cdc的tableAPI创建表，并指定从指定时间戳读取，只指定参数‘scan.startup.timestamp-millis’而不指定参数‘scan.startup.mode’，运行可以通过，并且使用的是默认的‘initial’模式。
（Use the cdc table API to create a table and specify to read from the specified timestamp. Only specify the parameter 'scan. startup. timestamp millis' and not the parameter 'scan. startup. mode'. It can be run and the default 'initial' mode is used.）

**Environment :**
 - Flink version :  1.16.0
 - Flink CDC version: 2.3.0
 - Database and version: mysql 8

**To Reproduce**
1、使用mysql-cdc创建一个表（create a table use mysql-cdc）
image

2、查询表（select the table）
一般来说，上述两个参数应该配合使用，如果缺乏其中一个，运行时都应该报错误，但是运行通过了。
（Generally speaking, the above two parameters should be used together. If one of them is missing, the runtime should get an error, but the operation passes.）
image

**Additional Description**

### PR DESCRIPTION
**Pull request by gtk96**
fix #1743

### CODE REVIEW

**Comment by rookiegao** [comment]
add validateTimeStamp and validateInitial verify whether the parameter is legal, It is obviously not enough to just add validateTimeStamp

We should double check the special startup mode and corresponding parameters to ensure that both are indispensable

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 1]
```diff
@@ -244,6 +246,34 @@ private static void validateSpecificOffset(ReadableConfig config) {
...
+        Optional<Long> timeStamp = config.getOptional(SCAN_STARTUP_TIMESTAMP_MILLIS);
+        if (!timeStamp.isPresent()) {
+            throw new ValidationException(
+                    String.format(
+                            "Unable to find a valid timestamp. %s are required.",
+                            SCAN_STARTUP_TIMESTAMP_MILLIS.key()));
+        }
+    }
+
+    private static void validateInitial(ReadableConfig config) {
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:259]: Do we also need this check for both latest and earliest?

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 2]
```diff
@@ -217,6 +218,7 @@ private static StartupOptions getStartupOptions(ReadableConfig config) {
                 return getSpecificOffset(config);
 
             case SCAN_STARTUP_MODE_VALUE_TIMESTAMP:
+                validateTimeStamp(config);
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:221]: validateTimestampMs

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 3]
```diff
@@ -244,6 +246,34 @@ private static void validateSpecificOffset(ReadableConfig config) {
         }
     }
 
+    private static void validateTimeStamp(ReadableConfig config) {
+        Optional<Long> timeStamp = config.getOptional(SCAN_STARTUP_TIMESTAMP_MILLIS);
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:250]: timestampMs

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 4]
```diff
@@ -244,6 +246,34 @@ private static void validateSpecificOffset(ReadableConfig config) {
...
+        if (!timeStamp.isPresent()) {
+            throw new ValidationException(
+                    String.format(
+                            "Unable to find a valid timestamp. %s are required.",
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:254]: String.format("Option %s are required when setting the option  %s to %s.", SCAN_STARTUP_TIMESTAMP_MILLIS.key(), SCAN_STARTUP_MODE.key(), SCAN_STARTUP_MODE_VALUE_TIMESTAMP)

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 5]
```diff
@@ -244,6 +246,34 @@ private static void validateSpecificOffset(ReadableConfig config) {
...
+                            SCAN_STARTUP_TIMESTAMP_MILLIS.key()));
+        }
+    }
+
+    private static void validateInitial(ReadableConfig config) {
+        Optional<String> gtidSet = config.getOptional(SCAN_STARTUP_SPECIFIC_OFFSET_GTID_SET);
+        Optional<String> binlogFilename = config.getOptional(SCAN_STARTUP_SPECIFIC_OFFSET_FILE);
+        Optional<Long> binlogPosition = config.getOptional(SCAN_STARTUP_SPECIFIC_OFFSET_POS);
+        Optional<Long> timeStamp = config.getOptional(SCAN_STARTUP_TIMESTAMP_MILLIS);
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:263]: timestampMs

**Review by ruanhang1993 — COMMENTED** [review]
Thanks for contributing @gaotingkai . I left some comments.

**Comment by gaotingkai** [comment]
@ruanhang1993 Do you have time to review it again? Thank you

**Comment by ruanhang1993** [comment]
@gtk96 Would you mind rebase the master and make the CI pass? Thanks~

**Comment by gtk96** [comment]
> @gtk96 Would you mind rebase the master and make the CI pass? Thanks~

done

**Comment by gtk96** [comment]
@ruanhang1993 If you have time, you can take a look for me.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 6]
```diff
@@ -259,6 +263,74 @@ private static void validateSpecificOffset(ReadableConfig config) {
         }
     }
 
+    private static void validateTimestampMs(ReadableConfig config) {
+        Optional<Long> timeStampMs = config.getOptional(SCAN_STARTUP_TIMESTAMP_MILLIS);
+        if (!timeStampMs.isPresent()) {
+            throw new ValidationException(
+                    String.format(
+                            "Option %s are required when setting the option %s to %s.",
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:271]: Option %s is required for the %s startup mode.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 7]
```diff
@@ -259,6 +263,74 @@ private static void validateSpecificOffset(ReadableConfig config) {
         }
     }
 
+    private static void validateTimestampMs(ReadableConfig config) {
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:266]: validateTimestampStartUpMode

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 8]
```diff
@@ -259,6 +263,74 @@ private static void validateSpecificOffset(ReadableConfig config) {
...
+            throw new ValidationException(
+                    String.format(
+                            "Option %s are required when setting the option %s to %s.",
+                            SCAN_STARTUP_TIMESTAMP_MILLIS.key(),
+                            SCAN_STARTUP_MODE.key(),
+                            SCAN_STARTUP_MODE_VALUE_TIMESTAMP));
+        }
+    }
+
+    private static void validateInitial(ReadableConfig config) {
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:278]: delete this method. Merge initial / earliest / latest into one.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 9]
```diff
@@ -259,6 +263,74 @@ private static void validateSpecificOffset(ReadableConfig config) {
...
+                    String.format(
+                            "Illegal parameter found. Either %s, %s, or %s and %s not required in initial mode.",
+                            SCAN_STARTUP_SPECIFIC_OFFSET_GTID_SET.key(),
+                            SCAN_STARTUP_TIMESTAMP_MILLIS.key(),
+                            SCAN_STARTUP_SPECIFIC_OFFSET_FILE.key(),
+                            SCAN_STARTUP_SPECIFIC_OFFSET_POS.key()));
+        }
+    }
+
+    private static void validateLatest(ReadableConfig config) {
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:296]: Merge this with initial / earliest check.
```java
validateStartUpMode(ReadableConfig config, String startupMode)
....

Options '%s', '%s', '%s' and '%s' must not be used for the %s startup mode.

```

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 10]
```diff
@@ -259,6 +263,74 @@ private static void validateSpecificOffset(ReadableConfig config) {
...
+
+        Optional<String> gtidSet = config.getOptional(SCAN_STARTUP_SPECIFIC_OFFSET_GTID_SET);
+        Optional<String> binlogFilename = config.getOptional(SCAN_STARTUP_SPECIFIC_OFFSET_FILE);
+        Optional<Long> binlogPosition = config.getOptional(SCAN_STARTUP_SPECIFIC_OFFSET_POS);
+        Optional<Long> timestampMs = config.getOptional(SCAN_STARTUP_TIMESTAMP_MILLIS);
+        if (gtidSet.isPresent()
+                || (binlogFilename.isPresent() && binlogPosition.isPresent())
+                || timestampMs.isPresent()) {
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:304]: gtidSet.isPresent() || binlogFilename.isPresent() || binlogPosition.isPresent() || timestampMs.isPresent()

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 11]
```diff
@@ -259,6 +263,74 @@ private static void validateSpecificOffset(ReadableConfig config) {
...
+                    String.format(
+                            "Illegal parameter found. Either %s, %s, or %s and %s not required in latest mode.",
+                            SCAN_STARTUP_SPECIFIC_OFFSET_GTID_SET.key(),
+                            SCAN_STARTUP_TIMESTAMP_MILLIS.key(),
+                            SCAN_STARTUP_SPECIFIC_OFFSET_FILE.key(),
+                            SCAN_STARTUP_SPECIFIC_OFFSET_POS.key()));
+        }
+    }
+
+    private static void validateEarliest(ReadableConfig config) {
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:315]: delete this method. Merge initial / earliest / latest into one.

**Review by ruanhang1993 — COMMENTED** [review]
Would you mind adding some tests ?

**Comment by gtk96** [comment]
The test will be added later.

**Comment by gtk96** [comment]
@ruanhang1993 If you have time, please review it for me.

**Pull request closed without merging**